cobra/util/solver.py

```diff
@@ -26,6 +29,108 @@ class SolverNotFound(Exception):
 """
 
 
+def linear_reaction_coefficients(model, reactions=None):
+    """Coefficient for the reactions in a linear objective
+
+    Parameters
+    ----------
+    model: cobra model
+        the model object that defined the objective
+    reactions: list
+        an optional list for the reactions to get the coefficients for. All
+        reactions if left missing.
+
+    Returns
+    -------
+    dict
+        A dictionary where the key is the reaction object and the value is
+        the corresponding coefficient. Empty dictionary if there are no
+        linear terms in the objective.
+    """
+    linear_coefficients = {}
+    reactions = model.reactions if not reactions else reactions
+    try:
+        objective_expression = model.solver.objective.expression
+        coefficients = objective_expression.as_coefficients_dict()
+    except AttributeError:
+        return linear_coefficients
+    for rxn in reactions:
+        forward_coefficient = coefficients.get(rxn.forward_variable, 0)
+        reverse_coefficient = coefficients.get(rxn.reverse_variable, 0)
+        if forward_coefficient != 0:
+            if forward_coefficient == -reverse_coefficient:
+                linear_coefficients[rxn] = float(forward_coefficient)
+    return linear_coefficients
+
+
+def set_objective(model, value, additive=False):
+    """ Set the model objective
+
+    Parameters
+    ----------
+    model: cobra model
+       The model to set the objective for
+    value: model.solver.interface.Objective such as
+           optlang.glpk_interface.Objective, sympy.Basic, dict, int,
+           list, string or Reaction
+
+        If the model objective is linear, the value can be a new Objective
+        object directly, a sympy expression which is used to create a new
+        Objective, a dictionary of linear coefficients where each key is a
+        reaction and the element the new coefficient (float), a list of
+        reactions where each reaction is interpreted to be meant to have an
+        objective coefficient of 1, a string interpreted as a single
+        reaction identifier or directly a Reaction object.
+
+        If the objective is not linear, only values of class Objective or
+        sympy expression's are allowed.
+    additive: bool
+        If true, add the terms to the current objective, otherwise start with
+        an empty objective
+    """
+    by_expression = (isinstance(value, sympy.Basic) or
+                     isinstance(value, model.solver.interface.Objective))
+    by_coefficient = isinstance(value, (dict, int, list, six.string_types,
+                                        cobra.Reaction))
+    not_supported = (
+        additive and (not model.objective.is_Linear and not by_expression))
```

Comment:
It kind of works since optlang keeps linear and quadratic coefficients apart. So setting linear coefficients in a mixed objective will just add those to the entire objective.

For instance, the following works:

```Python
In [28]: mod = create_test_model("textbook")

In [29]: mod.solver = "cplex"

In [30]: mod.objective = mod.solver.variables.ATPM**2

In [31]: mod.solver.objective.expression
Out[31]: 1.0*ATPM**2

In [32]: mod.solver.objective.set_linear_coefficients({mod.solver.variables.ATPM: 2})

In [33]: mod.solver.objective.expression
Out[33]: 1.0*ATPM**2 + 2.0*ATPM
```